### ISSUE TITLE
Fix tests

### ISSUE THREAD

**Issue by r0x0r** [r0x0r/pywebview#128]
Both Travis and AppVeyor tests fail at the moment. Travis seems to have a configuration issue. AppVeyor fails are semi-random and seem to do with race conditions. I could not reproduce AppVeyor failures locally.

**Comment by shivaprsdv** [comment]
Tests are now passing on both. Seems to be the result of fbe56e00b41d34cebcb6c489b65a5771d564fa6c.
Or is the issue solved?

**Comment by r0x0r** [comment]
Disabling 32-bit builds certainly helped, but there are still random fails from time to time. Still no idea what causes fails, as I could not reproduce this issue locally. Might be something fundamentally wrong how tests are run, for all I know.

**Comment by shivaprsdv** [comment]
Another thing, all tests are failing locally on macOS Sierra. On googling, it looks
likely an OS issue. Still thought I'd report.

**Comment by shivaprsdv** [comment]
Currently OSX builds on Travis are failing with a `sudo: pip: command not found`
error. This seems to be different from previous issues. Can we check if it is a bug
from their side?

**Comment by r0x0r** [comment]
Looks like their OSX configuration changed and pip is no longer on the path. I will investigate.

**Comment by shivaprsdv** [comment]
- [ ] [macOS] Fix mouse helper function

**Comment by r0x0r** [comment]
It should be noted that the mouse helper is sort of a hack. The reason it exists is because Coca events do not fire unless mouse is moved. I don't know why Cocoa has this behaviour and whether there is a better solution than moving a mouse cursor programmatically, but at least it had got the job done until now (I blame a recent OSX update for this).

**Comment by r0x0r** [comment]
The current GTK implementation is broken on master and tests do not catch that. It throws an exception in the GUI thread, which is reported to the console, but is not caught by the catch block in the test runner. I fixed the bug in `set_title` branch, test runner is yet to be addressed.

**Comment by shivaprsdv** [comment]
> I blame a recent OSX update for this
Tests are running fine for me locally (macOS High Sierra 10.13.2, built in Python 2.7)
Which update is causing the issue? Or is it a Travis only issue?

**Comment by r0x0r** [comment]
I think it might be a (High) Sierra issue. Anyhow I downgraded the OSX image on Travis to 10.11 and suddenly the tests pass. Let's keep it this way for the time being.

**Comment by shivaprsdv** [comment]
I think the `assert_js` function is badly formulated. It's leading to infinite loops
on Linux.

**Comment by r0x0r** [comment]
GTK or QT?

**Comment by shivaprsdv** [comment]
Both. I suspect the `while` loop in `assert_js` for the test failures on Travis. I
too am unable to reproduce the fails locally, but as it's only the js_bridge tests
that get stuck on the server, it could well be some race condition issue.

**Comment by r0x0r** [comment]
I managed to reproduce the issue a couple of times locally in the `evaluate-js-renovation` branch. The result of evaluate_js gets evaluated to None for some reason, which causes an endless loop. Still no idea what causes it.

**Comment by r0x0r** [comment]
What I have discovered is that evaluate_js implementations are not thread safe in QT and GTK. Sharing a single js_result object messes up results of multiple evaluate_js. I fixed the issue on both platforms, PyQt4 still has a problem with passing tests though. 
The GTK implementation locks evaluate_js, because of` document.title` serving as a shared object In other words multiple evaluate_js cannot be run simultaneously. This can be remedied, for example, by storing data of multiple evaluate_js calls in `document.title` as stringified JSON or something like that. Let's keep it as it is now, unless it causes performance problems.
